WIP: State Reconstructor

inxilpro opened the pull request:
Just posting this as a draft for discussion.

inxilpro committed:
wip

inxilpro committed:
Fix styling

inxilpro committed:
wip

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

inxilpro committed:
Allow EventStateRegistry to reset

inxilpro committed:
Fix styling

inxilpro committed:
Undo infinite loop

inxilpro committed:
GREEN

Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
refactoring

Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

inxilpro committed:
One more failing test

Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
Make test moar bad

Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
Fix styling

inxilpro committed:
wip

Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

inxilpro committed:
wip

inxilpro committed:
wip

inxilpro committed:
WIP

inxilpro committed:
wip

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
Fix styling

inxilpro committed:
Move singleton status to the state class

inxilpro committed:
A little bit of docs

inxilpro committed:
Fix styling

inxilpro committed:
Merge branch 'better-singletons' into state-reconstructor

inxilpro committed:
Start to refactor

inxilpro committed:
Fix styling

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

inxilpro committed:
wip

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro commented on src/Lifecycle/AggregateStateSummary.php:
```diff
@@ -0,0 +1,87 @@
+use Thunk\Verbs\State;
+use Thunk\Verbs\Support\StateIdentity;
+
+class AggregateStateSummary
```
This name is… obtuse. But the basic point of this class is to find ALL the States and Events that are related to a given set of starting States. We need this because if we're reconstituting state, and it relies on some other related state, we need that state to be in sync.

skylerkatz replied:
```php
class FindAllTheThings
{
    public static function leeroyyyyyyJenkins(State ...$states): static
}
```

inxilpro commented on src/Lifecycle/StateManager.php:
```diff
@@ -188,18 +200,47 @@ protected function loadMany(iterable $ids, string $type): StateCollection
+        // reconstituting in a recursive call, the root call is responsible for applying
+        // events, so we should also skip in that case.
+
+        if (! $this->is_replaying && ! $this->is_reconstituting) {
```
It might be useful to combine `is_replaying` and `is_reconstituting` into a single concept that means "should the state manager try to reconstitute state or not" — I don't know that the distinction matters inside the state manager.

inxilpro commented on src/Lifecycle/StateManager.php:
```diff
@@ -188,18 +200,47 @@ protected function loadMany(iterable $ids, string $type): StateCollection
+            try {
+                $this->is_reconstituting = true;
+
+                $summary = $this->events->summarize($state);
```
This probably should be able to take multiple states

inxilpro commented on src/Lifecycle/StateManager.php:
```diff
@@ -188,18 +200,47 @@ protected function loadMany(iterable $ids, string $type): StateCollection
+                $summary = $this->events->summarize($state);
+
+                // FIXME: We probably need to re-write all the snapshots after we're done
+                // FIXME: Swap out existing state manager, push all related states into new state manager
+                // FIXME: run all the event on them, swap them out
```
This is the correct approach, I think. Ultimately, we just need an isolated state manager to handle everything and then pull those values over into the global state manager after it's done.

inxilpro commented on src/Lifecycle/StateManager.php:
```diff
@@ -188,18 +200,47 @@ protected function loadMany(iterable $ids, string $type): StateCollection
+                            ->map(fn (State $state) => $state->last_event_id)
+                            ->filter();
+
+                        $min = $last_event_ids->min() ?? PHP_INT_MIN;
+                        $max = $last_event_ids->max() ?? PHP_INT_MIN;
+
+                        // If all states have had this or future events applied, just ignore them
+                        if ($min >= $event->id && $max >= $event->id) {
+                            return false;
+                        }
+
+                        // We should never be in a situation where some events are ahead and
+                        // others are behind, so if that's the case we'll throw an exception
+                        if ($max > $event->id && $min <= $event->id) {
+                            throw new RuntimeException('Trying to apply an event to states that are out of sync.');
+                        }
```
This is all to account for cases where one state is ahead of another. We may want to decide that in the beginning we just don't support that…?

inxilpro commented on src/Support/StateIdentity.php:
```diff
@@ -0,0 +1,35 @@
+use InvalidArgumentException;
+use Thunk\Verbs\State;
+
+class StateIdentity
```
There are too many different things that can point to a state. This class exists to normalize them all into one thing for ease of use around the codebase.

inxilpro commented on src/Testing/SnapshotStoreFake.php:
```diff
@@ -60,6 +60,21 @@ public function reset(): bool
+        foreach ($this->states as $type => $states) {
+            foreach ($states as $id => $state) {
+                if (in_array($id, $ids)) {
+                    uniqid($this->states[$type][$id]);
```
lol, I think this should be `unset` not `uniqid`

inxilpro commented on tests/Unit/StateReconstitutionTest.php:
```diff
@@ -0,0 +1,200 @@
+use Thunk\Verbs\Models\VerbSnapshot;
+use Thunk\Verbs\State;
+
+/*
+ * The Problem(s)
+ *
+ * FIRST PROBLEM:
+ * - We try to load state1, but we don't have an up-to-date snapshot
+ * - StateManager::load tries to reconstitute state from events
+ * - One of those Event::apply methods load state2
+ * - Best case scenario: we reconstitute state2 before continuing
+ * - Worst case scenario: reconstituting state2 tries to reconstitute state1, and we're in an infinite loop
+ * - (if no loop) state1 continues to reconstitute, but it's acting with state2 FULLY up-to-date, not
+ *   just up-to-date with where state1 happens to be
+ *
+ * TO TEST FIRST PROBLEM:
+ * - Event1 adds State1::counter to State2::counter and increments State2::counter
+ * - Event2 increments State2::counter
+ *
+ * SECOND PROBLEM:
+ * - We try to load state1, but we don't have an up-to-date snapshot
+ * - StateManager::load tries to reconstitute state from events
+ * - One of those Event::apply methods requires state1 and state2, so we need to load state2
+ * - Reconstituting state2 re-runs the same apply method on state2 before also running it on state1
+ * - Double-apply happens
+ */
```
I think we should probably start here and review whether these are actually the problems and if so confirm that the tests effectively test them.

inxilpro reviewed (commented):
@DanielCoulbourne @skylerkatz I just did a pass on this and added some notes in prep for Monday.

skylerkatz commented on src/Support/StateIdentity.php:
```diff
@@ -0,0 +1,35 @@
+        $state_id = data_get($source, 'state_id');
+        $state_type = data_get($source, 'state_type');
+
+        if (is_int($state_id) && is_string($state_type)) {
```
```suggestion
        if (
            (
                \Illuminate\Support\Str::isUlid($state_id)
                || \Illuminate\Support\Str::isUuid($state_id)
                || is_int($state_id)
            )
            && is_string($state_type)
        ) {
```

This is formatted terribly, but I think we need to check for more than just an `int` for the `$state_id`

inxilpro committed:
wip

Co-Authored-By: Daniel Coulbourne <429010+DanielCoulbourne@users.noreply.github.com>
Co-Authored-By: Skyler Katz <skylerkatz@hey.com>

inxilpro committed:
wip

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
wip

inxilpro committed:
wip

inxilpro committed:
wip

inxilpro committed:
wip

inxilpro committed:
wip

DanielCoulbourne committed:
WIP

DanielCoulbourne committed:
Fix styling

inxilpro committed:
Rename "Timeline" to "Replay"

inxilpro committed:
Merge branch 'state-reconstructor' of https://github.com/hirethunk/verbs into state-reconstructor

DanielCoulbourne committed:
WIP

DanielCoulbourne committed:
Fix styling

inxilpro committed:
Change order of arguments in state manager methods

inxilpro committed:
wip

inxilpro committed:
wip

inxilpro committed:
Move things around

inxilpro committed:
ugh

inxilpro committed:
Add replay & state reconstitution plan

inxilpro committed:
Rename StateManager to Scope, implement lazy state reconstitution

- Rename StateManager class to Scope and update all references
- Create ReconstitutingScope that reconstitutes stale states on-demand by replaying their connected component of events
- Add Scope::run() method for managing the current scope binding in the container
- Add Scope::singleton(), all(), willPrune(), and prune() methods
- Enable snapshot writing and pruning in Broker by implementing writeSnapshots() helper
- Add StoresSnapshots dependency to Broker
- Update cache contracts with values() method and prune-related methods
- Remove incomplete StateReconstructor and ReconstitutingStateManager classes
- Bind ReconstitutingScope in service provider
- Update Replay to use Scope::run() for scope management
- Rename StateManagerTest to ScopeTest and update test names
- Fix Date::setTestNow() calls to include timestamps
- Remove debug dump() statements from tests

inxilpro committed:
Fix state eviction during replay, singleton staleness detection, and harvest merging

- Restore snapshot hydration on the replay path (skip reconstitution but load from snapshots)
- Fix isStale() to match singletons by type only, mirroring event storage
- Fix harvest to merge into exact requested instances and never clobber live singletons
- Hoist setReplaying(true) outside the replay loop
- Add ReplayEvictionTest to pin snapshot reload during eviction

inxilpro committed:
Fix singleton state reconstitution: discovery and harvest keying

inxilpro committed:
Revert name change

inxilpro committed:
Fix styling

inxilpro committed:
Implement state pinning and refactor cache layer architecture

Pin states with in-flight events to prevent eviction during batch operations. Refactor MultiCache to support layered reads with backfill and separate read/write-only cache wrappers. Split authorize and validate into distinct phases. Fix StateIdentity to accept string IDs from database drivers. Add comprehensive cache and state identity tests.

inxilpro committed:
Merge branch 'main' into state-reconstructor

inxilpro committed:
Add .clave.local.json to .gitignore
